pkg/kubelet/kuberuntime/kuberuntime_manager.go

```diff
@@ -447,19 +457,46 @@ func (m *kubeGenericRuntimeManager) computePodContainerChanges(pod *v1.Pod, podS
 	// check the status of containers.
 	for index, container := range pod.Spec.Containers {
 		containerStatus := podStatus.FindContainerStatusByName(container.Name)
-		if containerStatus == nil || containerStatus.State != kubecontainer.ContainerStateRunning {
+		containerRef, err := kubecontainer.GenerateContainerRef(pod, &container)
+		if err != nil {
+			glog.Errorf("Couldn't make a ref to pod %q, container %v: '%v'", format.Pod(pod), container.Name, err)
+		}
+
+		// Was this container running the last time we checked?
+		wasRunning := false
+		if existingStatus != nil {
+			previousContainerStatus := kubecontainer.GetContainerStatus(existingStatus, container.Name)
+			if previousContainerStatus != nil && previousContainerStatus.State.Running != nil {
+				wasRunning = true
+			}
+		}
+
+		if containerStatus == nil {
+			// No container exists by this name, so just start it
+			changes.ContainersToStart[index] = "Starting container"
+			continue
+		} else if containerStatus.State != kubecontainer.ContainerStateRunning {
+			if wasRunning {
+				// The container was running but currently is not. Record that it exited.
+				exitMessage := fmt.Sprintf("Container exited with status %d", containerStatus.ExitCode)
+				if containerStatus.ExitCode == 0 {
+					m.recorder.Eventf(containerRef, v1.EventTypeNormal, events.ExitSuccess, exitMessage)
+				} else {
+					m.recorder.Eventf(containerRef, v1.EventTypeWarning, events.ExitFailure, exitMessage)
+				}
+			}
+
 			if kubecontainer.ShouldContainerBeRestarted(&container, pod, podStatus) {
-				message := fmt.Sprintf("Container %+v is dead, but RestartPolicy says that we should restart it.", container)
-				glog.Info(message)
-				changes.ContainersToStart[index] = message
+				glog.Infof("Container %+v is dead, but RestartPolicy says that we should restart it.", container)
```

Comment:
I don't think it's a warning from kubelet's perspective, it's normal that kubelet restarts containers.  It's abnormal from the user's perspective (ie. the person who's running their app), but that's what events are for to begin with IMO.